nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,573 @@
+
+  transform_tolerance_ = tf2::durationFromSec(transform_tolerance);
+  control_duration_ = 1.0 / control_frequency;
+
+  global_pub_ = node->create_publisher<nav_msgs::msg::Path>("received_global_plan", 1);
+  carrot_pub_ = node->create_publisher<geometry_msgs::msg::PointStamped>("lookahead_point", 1);
+  carrot_arc_pub_ = node->create_publisher<nav_msgs::msg::Path>("lookahead_collision_arc", 1);
+}
+
+void RegulatedPurePursuitController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type"
+    " regulated_pure_pursuit_controller::RegulatedPurePursuitController",
+    plugin_name_.c_str());
+  global_pub_.reset();
+  carrot_pub_.reset();
+  carrot_arc_pub_.reset();
+}
+
+void RegulatedPurePursuitController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type "
+    "regulated_pure_pursuit_controller::RegulatedPurePursuitController\"  %s",
+    plugin_name_.c_str());
+  global_pub_->on_activate();
+  carrot_pub_->on_activate();
+  carrot_arc_pub_->on_activate();
+}
+
+void RegulatedPurePursuitController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type "
+    "regulated_pure_pursuit_controller::RegulatedPurePursuitController\"  %s",
+    plugin_name_.c_str());
+  global_pub_->on_deactivate();
+  carrot_pub_->on_deactivate();
+  carrot_arc_pub_->on_deactivate();
+}
+
+double RegulatedPurePursuitController::getLookAheadDistance(const geometry_msgs::msg::Twist & speed)
+{
+  // If using velocity-scaled look ahead distances, find and clamp the dist
+  // Else, use the static look ahead distance
+  double lookahead_dist = 0.0;
+  if (use_velocity_scaled_lookahead_dist_) {
+    lookahead_dist = speed.linear.x * lookahead_time_;
+    lookahead_dist = std::clamp(lookahead_dist, min_lookahead_dist_, max_lookahead_dist_);
+  } else {
+    lookahead_dist = lookahead_dist_;
+  }
+
+  return lookahead_dist;
+}
+
+std::unique_ptr<geometry_msgs::msg::PointStamped> RegulatedPurePursuitController::createCarrotMsg(
+  const geometry_msgs::msg::PoseStamped & carrot_pose)
+{
+  auto carrot_msg = std::make_unique<geometry_msgs::msg::PointStamped>();
+  carrot_msg->header = carrot_pose.header;
+  carrot_msg->point.x = carrot_pose.pose.position.x;
+  carrot_msg->point.y = carrot_pose.pose.position.y;
+  carrot_msg->point.z = 0.01;  // publish right over map to stand out
+  return carrot_msg;
+}
+
+geometry_msgs::msg::TwistStamped RegulatedPurePursuitController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed)
+{
+  // Transform path to robot base frame
+  auto transformed_plan = transformGlobalPlan(pose);
+
+  // Find look ahead distance and point on path and publish
+  const double lookahead_dist = getLookAheadDistance(speed);
+  auto carrot_pose = getLookAheadMarker(lookahead_dist, transformed_plan);
+  carrot_pub_->publish(std::move(createCarrotMsg(carrot_pose)));
```

Comment:
Add parameter to enable or disable visualization? Though keeping it on permanently, is no harm as of now.

Reply:
There's no reason to, this is really small